### PR TITLE
8331553: Windows JVM leaks Event and Thread handles when multiple threads are used

### PR DESCRIPTION
**Pull request by djelinski**
We use 2 ParkEvent instances per thread. The ParkEvent objects are never freed, but they are recycled when a thread dies, so the number of live ParkEvent instances is proportional to the maximum number of threads that were live at any time.

On Windows, the ParkEvent object wraps a kernel Event object. Kernel objects are a limited and costly resource. In this PR, I replace the use of kernel events with user-space synchronization.

The new implementation uses WaitOnAddress and WakeByAddressSingle methods to implement synchronization. The methods are available since Windows 8. We only support Windows 10 and newer, so OS support should not be a problem.

WaitOnAddress was observed to return spuriously, so I added the necessary code to recalculate the timeout and continue waiting.

Tier1-5 tests passed. Performance tests were... inconclusive. For example, `ThreadOnSpinWaitProducerConsumer` reported 30% better results, while `LockUnlock.testContendedLock` results were 50% worse. 

Thoughts?

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8331553](https://bugs.openjdk.org/browse/JDK-8331553): Windows JVM leaks Event and Thread handles when multiple threads are used (**Bug** - P4)

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to [23b70220](https://git.openjdk.org/jdk/pull/19778/files/23b702209cb94e6adb9fa05f4a889527d2928df8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19778/head:pull/19778` \
`$ git checkout pull/19778`

Update a local copy of the PR: \
`$ git checkout pull/19778` \
`$ git pull https://git.openjdk.org/jdk.git pull/19778/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19778`

View PR using the GUI difftool: \
`$ git pr show -t 19778`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19778.diff">https://git.openjdk.org/jdk/pull/19778.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19778#issuecomment-2176979906)

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
This is interesting. This is not an API I had heard of before now. I'm curious about how it works under the covers (some kind of hashtable of watched addresses I assume) and what its performance and memory profile is like when large numbers of threads may be involved. Hopefully it is early enough in the release cycle to catch any issues.

Note that there are still some references to WaitForSingleObject in the comments.

If this is simply a "lightweight event" mechanism then we could replace the other Event usages on Windows ... future RFE's of course.

**Comment by dholmes-ora** [comment]
BTW found this: https://devblogs.microsoft.com/oldnewthing/20160823-00/?p=94145

As I expected it is just a hashtable so may have scaling limitations.

**Comment by tstuefe** [comment]
Interesting, did not know that.

Apart from the unknown performance characteristics, what about backward compatibility? Do we care for Windows older than 8?

I share @dholmes-ora concern about scalibility. Your patch replaces the underlying mechanics for ObjectMonitor, right? We can have a lot of OMs, depending on how fast deflation is. Not many of them would be contended though.

**Comment by djelinski** [comment]
As you found out already, the implementation is based on a hash table, so access will be slower with many threads waiting at the same time. The hash table is stored in user space (in PEB), and the implementation reportedly doesn't require any kernel resources.

I'm not sure what to do with the remaining reference to `WaitForSingleObject`; it explains why we decompose the timeout by pointing to `EventWait`, which no longer exists as far as I could tell. I'll search the history some more before I decide what to do with that comment.

I don't think we care about pre-Win8 OSes at this point. JDK-11 was the last version to support Windows 7, I guess we won't backport this change there.

The patch replaces the underlying mechanics of ObjectMonitor. The ParkEvent is only used when a thread is blocked on a monitor, so the number of concurrent waits will never be greater than the total number of running threads.

I'll check if I can run Renaissance philosophers. Could you explain how UseHeavyMonitors changes the benchmark?

**Comment by tstuefe** [comment]
> As you found out already, the implementation is based on a hash table, so access will be slower with many threads waiting at the same time. The hash table is stored in user space (in PEB), and the implementation reportedly doesn't require any kernel resources.
> 
> I'm not sure what to do with the remaining reference to `WaitForSingleObject`; it explains why we decompose the timeout by pointing to `EventWait`, which no longer exists as far as I could tell. I'll search the history some more before I decide what to do with that comment.
> 
> I don't think we care about pre-Win8 OSes at this point. JDK-11 was the last version to support Windows 7, I guess we won't backport this change there.
> 
> The patch replaces the underlying mechanics of ObjectMonitor. The ParkEvent is only used when a thread is blocked on a monitor, so the number of concurrent waits will never be greater than the total number of running threads.
> 
> I'll check if I can run Renaissance philosophers. Could you explain how UseHeavyMonitors changes the benchmark?

With UseHeavyMonitors, even uncontended locks are inflated to OMs. Otherwise, we use stack locking resp. now lightweight locking. The benchmark itself generates a lot of locks IIRC.

**Comment by djelinski** [comment]
> With UseHeavyMonitors, even uncontended locks are inflated to OMs.

I don't think that matters much for the changed code; only the implementation of park/unpark methods changed, and these methods are called for contended locks only.

I ran the philosophers benchmark without heavy monitors; the results before and after were very similar, the difference between the averages was way below the standard deviation. I'll look into running the benchmark with heavy monitors next.

I found the EventWait code deep in the archives...
```
  // On Windows millis values greater than max_millis_value
  // have special semantics.
  const DWORD max_millis_value = 0xFFFF0000;
```
searching the net for FFFF0000, I found another piece of code [here](https://opensource.apple.com/source/llvmgcc42/llvmgcc42-2118/gcc/ada/s-taprop-mingw.adb.auto.html):
```
      Time_Out_Max : constant DWORD := 16#FFFF0000#;
      --  NT 4 cannot handle timeout values that are too large,
      --  e.g. DWORD'Last - 1
```
I will update the comment.

File: src/hotspot/os/windows/os_windows.cpp [hunk 1]
```diff
@@ -5582,37 +5548,34 @@ int PlatformEvent::park(jlong Millis) {
...
+  jlong waitStart = os::javaTimeNanos();
+  // preserve initial millis for remaining time calculation on spurious wakeup
+  jlong initialMillis = Millis;
+  BOOL rv;
+  while ((v = _Event) < 0 && Millis > 0) {
+    DWORD prd = Millis;     // set prd = MIN (Millis, MAXTIMEOUT)
     if (Millis > MAXTIMEOUT) {
       prd = MAXTIMEOUT;
     }
     HighResolutionInterval *phri = nullptr;
```
Comment (viktorklang-ora) [inline src/hotspot/os/windows/os_windows.cpp:5565]: @djelinski Is this even used?

Comment (djelinski) [reply src/hotspot/os/windows/os_windows.cpp:5565]: yeah, it's the C++ construction where the constructor and the destructor have side effects. It increases the system timer resolution, unless `ForceTimeHighResolution` is set. `ForceTimeHighResolution`, contrary to its name and help text, forces [low timer resolution](https://bugs.openjdk.org/browse/JDK-6435126). Sigh.

Comment (viktorklang-ora) [reply src/hotspot/os/windows/os_windows.cpp:5565]: Oh. That's.... interesting. 😞

Comment (tstuefe) [reply src/hotspot/os/windows/os_windows.cpp:5565]: Of course pre-existing, but the typical pattern we use with RAII objects that may or may not do something is to give it a constructor argument, e.g. `bool activate`, that controls if it is activated. Allocating an RAII object with new/delete seems odd.

**Comment by tstuefe** [comment]
What I don't understand: 

Events are kernel events. We can release Events. WaitOnAddress builds up a kernel-side hash table, but is there a way to ever release these addresses and shrink the table, if you don't need the synchronization anymore? Can you ever release the underlying memory?

**Comment by djelinski** [comment]
I don't fully understand it either, but:
- the table is in user space, not in the kernel
- the entries are only inserted when a wait is active, and are removed when the wait is done
- when there's a race in waking up the thread, the kernel remembers that the thread needs to be woken up immediately the next time it waits; I suppose that's a flag in the thread structure, no extra memory required, but couldn't find a definitive answer in the docs.

Sources:
- Windows internals book https://www.oreilly.com/library/view/windows-internals-part/9780135462348/ch08.xhtml
- Raymond Chen's blog: https://devblogs.microsoft.com/oldnewthing/20160826-00/?p=94185

**Comment by djelinski** [comment]
the GHA failure looks related, but I can't reproduce it locally. Will dig...

File: src/hotspot/os/windows/os_windows.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/os/windows/os_windows.cpp:5565]: Please just let this sleeping dog lie. ;-)

**Comment by dholmes-ora** [comment]
@djelinski why did you close the PR??

**Comment by djelinski** [comment]
Need some time to experiment with the gha failure. I hope to reopen when I figure it out.

Looks like either the synchronization is so much slower now, or there's a bug hiding somewhere.

**Comment by dholmes-ora** [comment]
Okay - could have moved back to Draft. Not sure it makes any practical difference.

**Comment by dholmes-ora** [comment]
> the GHA failure looks related, but I can't reproduce it locally. Will dig...

The failure in jdk/jshell/ToolTabSnippetTest.java? That test is somewhat flakey - search JBS.

**Comment by djelinski** [comment]
Yes, that test. It failed with a timeout in 3 out of 3 runs. The test is doing producer/consumer with Object.wait/notifyAll, so it's most likely related.

**Comment by djelinski** [comment]
so... it appears that the ToolTabSnippetTest failure is related to a change in timing caused by the change in synchronization. Before my PR, the test takes anywhere between 10-40 seconds to complete. After my PR, on the GHA machines it takes 23 minutes to complete. The slowdown is observable on Win 2016 and Win 2019, Win 2022 and the client systems (10 and 11) still finish under 30 seconds.

The problem is related to the structure of this test; the test implements a producer/consumer pattern using Object.wait/notifyAll. The producer thread pushes text to a StringBuilder one byte at a time. The consumer thread runs a regex match on the StringBuilder contents. With the original implementation, the consumer thread only woke once every few hundred bytes, so the cost of the regex matching was reasonable. With the new implementation, it wakes up much more frequently, and the regex execution cost causes the test run time explosion.

**Comment by dholmes-ora** [comment]
>  With the new implementation, it wakes up much more frequently, 

Why? The number of unparks is unchanged. Are we seeing spurious unparks that are consuming too much CPU?

**Comment by djelinski** [comment]
well notifyAll doesn't give control to the waiting thread; the waiting thread wakes up, but still needs to compete with the notifying thread for the monitor. With WaitOnAddress, the waiting thread wakes up and acquires the monitor much faster than with the original event-based implementation.

**Comment by dholmes-ora** [comment]
> the waiting thread wakes up, but still needs to compete with the notifying thread for the monitor. 

No, a notified thread is moved from the wait-queue to the monitor-entry queue. It is only unparked once the notifier thread releases the monitor.

**Comment by dholmes-ora** [comment]
Okay but based on the change in code, this change in behaviour is really not making sense to me - unless we have lots of spurious wakeups.

**Comment by djelinski** [comment]
The test uses 2 threads that are serialized - when one is working, the other one is blocked, and vice versa. The publisher thread is publishing characters one at a time, so it's doing the same amount of work on every test run. The subscriber thread is doing a large amount of work every time it gets the monitor, and the amount of work does not depend on the number of bytes published since the last time the subscriber thread was run.

With the original code, the publisher thread was getting the monitor ~100x more often than the subscriber thread; with the new code, the publisher thread gets the monitor 2x more often than the subscriber thread. In other words, the subscriber thread is running 50x more often than before, and does proportionally more work. This translates the running time increase.

**Comment by dholmes-ora** [comment]
> With the original code, the publisher thread was getting the monitor ~100x more often than the subscriber thread; with the new code, the publisher thread gets the monitor 2x more often than the subscriber thread. 

Okay but still, based on the code change, I do not understand why this would be the case.

**Comment by djelinski** [comment]
WaitOnAddress is reported to spin before yielding. It might be related.

**Comment by viktorklang-ora** [comment]
@djelinski @dholmes-ora Couldn't it be that the producer thread is already scheduled, so it can reacquire the monitor whereas if the consumer was parked then the wake-up "lag" means that statistically it will lose?

**Comment by dholmes-ora** [comment]
@viktorklang-ora the question is why the two APIs seems to be behaving so differently. I would expect `WaitForSingleObject` and `WaitOnAddress` to be very similar internally - though speculation on my part.

**Comment by viktorklang-ora** [comment]
@dholmes-ora From the [documentation](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress) it would seem like WaitOnAddress behaves differently from WaitForSingleObject.

For instance, the following passage: "WaitOnAddress is more efficient [...] because WaitOnAddress does not interfere with the thread scheduler." <-- this suggests (to me) that it is some form of spin-wait.

**Comment by viktorklang-ora** [comment]
@dholmes-ora I guess it depends on one's definition of interference. :)

**Pull request closed without merging**